db/system_keyspace.cc

```diff
@@ -1811,6 +1811,95 @@ class nodetool_status_table : public memtable_filling_virtual_table {
     }
 };
 
+class describe_ring_table : public streaming_virtual_table {
+public:
+    describe_ring_table() : streaming_virtual_table(build_schema()) {
+        _shard_aware = true;
+    }
+
+    static schema_ptr build_schema() {
+        auto id = generate_legacy_id(NAME, "describe_ring");
+        return schema_builder(NAME, "describe_ring", std::make_optional(id))
+            .with_column("keyspace_name", utf8_type, column_kind::partition_key)
+            .with_column("start_token", utf8_type, column_kind::clustering_key)
+            .with_column("endpoint", inet_addr_type, column_kind::clustering_key)
+            .with_column("end_token", utf8_type)
+            .with_column("dc", utf8_type)
+            .with_column("rack", utf8_type)
+            .with_version(generate_schema_version(id))
+            .build();
+    }
+
+    dht::decorated_key make_partition_key(const sstring& name) {
+        return dht::decorate_key(*_s, partition_key::from_single_value(*_s, data_value(name).serialize_nonnull()));
+    }
+
+    clustering_key make_clustering_key(sstring start_token, gms::inet_address host) {
+        return clustering_key::from_exploded(*_s, {
+            data_value(start_token).serialize_nonnull(),
+            data_value(host).serialize_nonnull()
+        });
+    }
+
+    struct endpoint_details_cmp {
+        bool operator()(const dht::endpoint_details& l, const dht::endpoint_details& r) const {
+            return inet_addr_type->less(
+                data_value(gms::inet_address(l._host)).serialize_nonnull(),
+                data_value(gms::inet_address(r._host)).serialize_nonnull());
+        }
+    };
+
+    future<> execute(reader_permit permit, result_collector& result, const query_restrictions& qr) override {
+        return seastar::async([this, permit, &result, &qr] {
+            auto& ss = service::get_local_storage_service();
```

Comment:
This is filling the hole that @xemul is digging. You should receive the needed services via the constructor, not via globals that we are trying to remove.

Reply:
Done.